staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,570 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"k8s.io/kube-openapi/pkg/validation/strfmt"
+)
+
+// UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
+// The provided schema is expected to exactly match the fields we expose to CEL. I.e.
+// for resource roots, type meta and object meta schemas must be present.
+func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Structural) ref.Val {
+	if unstructured == nil {
+		if schema.Nullable {
+			return types.NullValue
+		}
+		return types.NewErr("invalid data, got null for schema with nullable=false")
+	}
+	if schema.XIntOrString {
+		switch unstructured.(type) {
+		case string, int64:
+			return types.DefaultTypeAdapter.NativeToValue(unstructured)
+		}
+		return types.NewErr("invalid data, expected XIntOrString value to be either a string or integer")
```

Comment:
Presumably, if the data doesn't match the declared type, the openapi validation was already going to throw up an error. How does this error surface to the user alongside that one? (same question goes for all the `invalid data, expected $type to match the provided schema with type=$type` errors returned from this function)

can be a follow-up (even for beta), but it would be good to have an issue or TODO to have a test that exercises all the errors returned from this as a user would encounter them submitting a CR... I suspect there would be a lot of duplication between the openapi failures and these, and some of the errors returned from this could be made more user-friendly (like I'm not sure returning the entire base64 string in `types.NewErr("Invalid byte formatted string %s: %v", str, err)` is what we want)

Reply:
Right. See my other answers. This is intended as a backstop, and if the error surfaces it will be a runtime CEL error and it means we've got a bug.

I've added tests cases for the invalid data types. The ones for invalid schemas are not as easy to write because there is so much other code  validating and expecting the schema structure to be valid that I can't trigger the cases easily.

Reply:
ok... good to know it's a backstop. also it's a good sign when error cases get harder and harder to trigger. for some reason I thought the CEL validation ran alongside the openapi validation even if there were errors from the openapi bits

Reply:
so... I played around with this, and it looks like these errors run and surface to the user when they provide data that doesn't match types: see https://gist.github.com/liggitt/5bb0ece637e560ec906ee6a0786eb3bd

Reply:
I still think this cleanup can be a follow-up, but I think there's work we should track for beta:
1. could make the messages more succinct or ideally lean on the openapi type checking and just return a "data was not of the expected type while evaluating rule: %s" or something
2. should probably use types that match the serialized API instead of go types (e.g. not `[]interface{}` or `map[string]interface{}`)

Reply:
Oh, you're right. I'll track this for follow up.

Reply:
https://github.com/kubernetes/kubernetes/issues/106440